Source/DafnyCore/AST/CompilationData.cs

```diff
@@ -5,7 +5,7 @@
 namespace Microsoft.Dafny;
 
 public class CompilationData {
-  public readonly FreshIdGenerator IdGenerator = new();
+  public readonly FreshIdGenerator IdGenerator = new CompilationIdGenerator();
```

Comment:
Refactoring `CompilationIdGenerator` to `CodeGenIdGenerator` done.
Now that you say it, I should also remove this line as it indicates a global id generator, which I'm removing away.

To answer your initial question, I want to avoid generating unique ID that are unique across the whole program. With such unique ID, any small change in a dfy file can potentially change all the generated cs files. This PR removes the use of global ID generators and scopes them to declarations, so that local changes in a dfy won't impact globally all other cs files, at least for temporary and local variable identifier numbers.

Therefore, I identified that there was a global static ID generator used to give a unique name to all local variables, and got rid of it, so that the name is unique only within one declaration.
Your remark helped me identify two other sources of global ID generation, which are ModuleResolver and MatchFlattener. I similarly removed both of them by stating that every code context (typically declarations) should have their own fresh id generator.

I also found that there were two kinds of ID generators, and we are mixing them. This PR also separates code generation ID generation from verification ID generation. The latter uses nested ID generation for variables that won't happen in the generated code.
There is still the possibility that the codeGenIdGenerator be used for matches in ghost code, so that the editing of ghost code might impact the fresh IDs, but I think this is a secondary issue that does not need to be addressed.


Let me know if there is still some shaded

Reply:
Thanks, great explanation. One thing about the Dafny codebase that confuses me is that variable such as locals or parameters are not considered declarations, even though I think outside of Dafny you'd call those declarations.

I pushed a refactoring that changes `formal.CompileName(SinglePassCodeGenerator.FormalIdGenerator)` to `formal.CompileName`, so the `FormalIdGenerator` is unneeded.